# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [4]:
import numpy as np
from scipy import stats

# Дані з умови
data = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7
n = len(data)
df = n - 1

# 2. Обчислення вибіркових статистик 
x_bar = np.mean(data)
s = np.std(data, ddof=1)

# 3. Виконання правостороннього t-тесту
t_stat, p_val = stats.ttest_1samp(data, mu0, alternative='greater')
alpha1 = 0.10
alpha2 = 0.05
t_crit1 = stats.t.ppf(1 - alpha1, df)
t_crit2 = stats.t.ppf(1 - alpha2, df)

print("Відповіді для пункту 1")
print(f"1. Гіпотези: H0: μ = 123.7, Ha: μ > 123.7 (правосторонній t-тест)")
print(f"2. Статистики: x_bar={x_bar:.2f}, s={s:.2f}, n={n}")
print(f"3. t-статистика={t_stat:.3f}, p-value={p_val:.4f}, t-критичне(0.10)={t_crit1:.3f}")


# 4. Прийняття рішення 
print(f"4. Рішення (при α=0.10):")
if p_val < alpha1:
    print("   Відхиляємо H0, оскільки p-value < 0.10.")
else:
    print("   Не відхиляємо H0, оскільки p-value >= 0.10.")

# 5. Висновок 
print("5. Висновок:")
if p_val < alpha1:
    print("   Так, редизайн сайту справді подовжив час до покупки.")
else:
    print("   Ні, нема достатньо доказів, що редизайн подовжив час.")

# 6. Зміна рівня значущості 
print(f"6. Зміна рівня значущості (α=0.05):")
if p_val < alpha2:
    print(f"   Рішення не змінилося: при α=0.05 ми також відхиляємо H0 (p-value {p_val:.4f} < 0.05).")
else:
    print(f"   Рішення змінилося: при α=0.05 ми не маємо достатньо доказів.")

Відповіді для пункту 1
1. Гіпотези: H0: μ = 123.7, Ha: μ > 123.7 (правосторонній t-тест)
2. Статистики: x_bar=132.29, s=10.98, n=7
3. t-статистика=2.069, p-value=0.0420, t-критичне(0.10)=1.440
4. Рішення (при α=0.10):
   Відхиляємо H0, оскільки p-value < 0.10.
5. Висновок:
   Так, редизайн сайту справді подовжив час до покупки.
6. Зміна рівня значущості (α=0.05):
   Рішення не змінилося: при α=0.05 ми також відхиляємо H0 (p-value 0.0420 < 0.05).


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [5]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

#Дані з умови задачі
nobs = 1200         # обсяг вибірки
count = 311         # кількість людей, що впізнали логотип
p0 = 0.23           # гіпотетична пропорція (H0)
alpha = 0.03        # рівень значущості

#1. Z-тест для пропорції (правосторонній, бо "більше ніж 23%")
stat, pval = proportions_ztest(count, nobs, value=p0, alternative='larger')

#2. Довірчий інтервал на рівні значущості alpha=0.03 (для довірчого інтервалу 1 - alpha = 0.97)
confint = proportion_confint(count, nobs, alpha=alpha, method='normal')

print("Тест для однієї пропорції (впізнаваність логотипу)")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")


Тест для однієї пропорції (впізнаваність логотипу)
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


Як інтерпретувати результати

1. Односторонній тест

   * $H_0: p = 0.23$
   * $H_1: p > 0.23$
   * Оскільки $p\text{-value} = 0.0106 < 0.03$, відхиляємо $H_0$.
   Є статистично значущі докази, що після рекламної кампанії частка дорослих, які впізнають логотип, більша за 23%.

2. 97% довірчий інтервал (0.2317, 0.2866)

   * Це означає, що з імовірністю 97% справжня частка лежить у межах **від 23.17% до 28.66%**.
   * Нижня межа інтервалу 23.17 лише рохи перевищує історичне значення ($23\%$). Якби ми проводили більш суворий двосторонній тест (де значення $\alpha$ розподіляється на обидва боки), ми могли б не отримати достатньо доказів для відхилення $H_0$.
   * Але в **односторонньому тесті** — ми вже отримуємо значущість.